controllers/etcd_copy_backups_controller.go

```diff
@@ -0,0 +1,446 @@
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/util/retry"
+	"k8s.io/utils/pointer"
+
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+)
+
+// EtcdReconciler reconciles a Etcd object
+type EtcdCopyBackupsTaskReconciler struct {
+	client.Client
+	Scheme       *runtime.Scheme
+	Config       *rest.Config
+	ImageVector  imagevector.ImageVector
+	chartApplier kubernetes.ChartApplier
+	logger       logr.Logger
+}
+
+// NewEtcdReconciler creates a new EtcdReconciler object
+func NewEtcdCopyBackupsTaskReconciler(mgr manager.Manager) (*EtcdCopyBackupsTaskReconciler, error) {
+	return (&EtcdCopyBackupsTaskReconciler{
+		Client: mgr.GetClient(),
+		Scheme: mgr.GetScheme(),
+		Config: mgr.GetConfig(),
+		logger: log.Log.WithName("etcd-backup-copy-controller"),
+	}).InitializeControllerWithChartApplier()
+}
+
+// NewEtcdReconcilerWithImageVector creates a new EtcdReconciler object
+func NewEtcdEtcdCopyBackupsTaskReconcilerWithImageVector(mgr manager.Manager) (*EtcdCopyBackupsTaskReconciler, error) {
+	ec, err := NewEtcdCopyBackupsTaskReconciler(mgr)
+	if err != nil {
+		return nil, err
+	}
+	return ec.InitializeControllerWithImageVector()
+}
+
+// InitializeControllerWithImageVector will use EtcdReconciler client to initialize image vector for etcd
+// and backup restore images.
+func (r *EtcdCopyBackupsTaskReconciler) InitializeControllerWithImageVector() (*EtcdCopyBackupsTaskReconciler, error) {
+	imageVector, err := imagevector.ReadGlobalImageVectorWithEnvOverride(getImageYAMLPath())
+	if err != nil {
+		return nil, err
+	}
+	r.ImageVector = imageVector
+	return r, nil
+}
+
+// Reconcile reconciles the EtcdCopyBackupsTask.
+func (r *EtcdCopyBackupsTaskReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	copyTask := &druidv1alpha1.EtcdCopyBackupsTask{}
+	if err := r.Get(ctx, req.NamespacedName, copyTask); err != nil {
+		if apierrors.IsNotFound(err) {
+			// Object not found, return. Created objects are automatically garbage collected.
+			// For additional cleanup logic use finalizers.
+			return ctrl.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return ctrl.Result{}, err
+	}
+
+	if !copyTask.DeletionTimestamp.IsZero() {
+		return r.delete(ctx, copyTask)
+	}
+	return r.reconcile(ctx, copyTask)
+}
+
+func (r *EtcdCopyBackupsTaskReconciler) reconcile(ctx context.Context, copyTask *druidv1alpha1.EtcdCopyBackupsTask) (ctrl.Result, error) {
+	logger := r.logger.WithValues("etcd", kutil.Key(copyTask.Namespace, copyTask.Name).String(), "operation", "reconcile")
+	logger.Info("Starting operation")
+
+	if err := r.addFinalizerToObject(ctx, copyTask, logger); err != nil {
+		return ctrl.Result{
+			Requeue: true,
+		}, err
```

Comment:
https://github.com/gardener/etcd-druid/pull/234#discussion_r714916480 We should either match the stile in the project or not.
Anyway, I don't mind to make all `return Results{}` to be in  one line one the whole project, because I don't like this either.